### PR TITLE
WF-752: Added routing examples for communication inside Journeys

### PR DESCRIPTION
**Pull request by manisha-backbase**
1. Added examples for route parameter, query parameters, navigationExtras state, fragments

### CODE REVIEW

File: libs/transactions/src/lib/views/transactions-view/transactions-view.component.ts [hunk 1]
```diff
@@ -27,11 +27,25 @@ export class TransactionsViewComponent {
...
     private readonly transactionsService: TransactionsHttpService,
     @Optional()
     @Inject(TRANSACTIONS_JOURNEY_COMMUNICATION_SERIVCE)
     private externalCommunicationService: TransactionsCommunicationService
   ) {}
+
+  ngOnInit(): void {
+    this.subscription.add(
+      this.route.queryParams.subscribe((params) => {
+        this.selectedAccount = params['id'];
```
Comment (kirillsud) [inline libs/transactions/src/lib/views/transactions-view/transactions-view.component.ts:43]: Why not the snapshot or `AsyncPipe`?

File: libs/transactions/src/lib/views/transaction-details/transaction-details.component.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+@Component({
+  templateUrl: './transaction-details.component.html',
+  selector: 'bb-transaction-details',
+})
+export class TransactionDetailsComponent implements OnInit {
+  public transactionDetails;
+  private routerState: NavigationExtras | undefined;
+  constructor(public route: ActivatedRoute, private router: Router) {
+    this.routerState = this?.router?.getCurrentNavigation()?.extras.state;
```
Comment (kirillsud) [inline libs/transactions/src/lib/views/transaction-details/transaction-details.component.ts:12]: The best practice is to resolve the data in the `Router.resolve` and transfer to the component only ready data

File: libs/transactions/src/lib/transactions-journey.module.ts [hunk 3]
```diff
@@ -17,21 +17,28 @@ import { TransactionsJourneyConfiguration } from './services/transactions-journe
 import { TransactionsRouteTitleResolverService } from './services/transactions-route-title-resolver.service';
 import { TransactionsHttpService } from './services/transactions.http.service';
 import { TransactionsViewComponent } from './views/transactions-view/transactions-view.component';
+import { TransactionDetailsComponent } from './views/transaction-details/transaction-details.component';
```
Comment (kirillsud) [inline libs/transactions/src/lib/transactions-journey.module.ts:20]: In GAS for the all view components we use `-view` suffix

File: libs/transactions/src/lib/transactions-journey.module.ts [hunk 4]
```diff
@@ -17,21 +17,28 @@ import { TransactionsJourneyConfiguration } from './services/transactions-journe
...
+      title: TRANSLATIONS.transactionsTitle,
+    },
+    resolve: {
+      title: TransactionsRouteTitleResolverService,
+    },
   },
-  resolve: {
-    title: TransactionsRouteTitleResolverService,
+  {
+    path: 'transaction-details/:id',
```
Comment (kirillsud) [inline libs/transactions/src/lib/transactions-journey.module.ts:38]: maybe just

```suggestion
    path: ':id',
```

File: libs/transactions/src/lib/transactions-journey.module.ts [hunk 5]
```diff
@@ -69,7 +77,7 @@ export class TransactionsJourneyModule {
     return {
       ngModule: TransactionsJourneyModule,
       providers: [
-        provideRoutes([route || defaultRoute]),
+        provideRoutes([...(defaultRoute || route)]),
```
Comment (kirillsud) [inline libs/transactions/src/lib/transactions-journey.module.ts:80]:
```suggestion
        provideRoutes([...(route || defaultRoute)]),
```

File: libs/transactions/src/lib/views/transactions-view/transactions-view.component.html [hunk 6]
```diff
@@ -9,9 +9,18 @@ <h1 *ngIf="title !== ''" data-role="transactions-view__title">
...
         <li
-          *ngFor="let item of items | filterTransactions: filter"
+          *ngFor="
+            let item of items | filterTransactions: filter:selectedAccount
+          "
           class="bb-list__item"
           data-role="transactions-view__item-container"
+          [routerLink]="['/transactions/transaction-details', item.id]"
+          [state]="{
+            queryParams: item,
```
Comment (kirillsud) [inline libs/transactions/src/lib/views/transactions-view/transactions-view.component.html:19]: We shouldn't recommend passing all data about the record in the URL. `id` should be enough and we can put the record data to the state, then check it on the route resolution stage and load the data by `id` if the state is empty.

File: libs/transactions/src/lib/views/transactions-view/transactions-view.component.html [hunk 7]
```diff
@@ -9,9 +9,18 @@ <h1 *ngIf="title !== ''" data-role="transactions-view__title">
...
+            skipLocationChange: false,
```
Comment (kirillsud) [inline libs/transactions/src/lib/views/transactions-view/transactions-view.component.html:20]: As I remember it is `false` by default

File: libs/transactions/src/lib/views/transactions-view/transactions-view.component.html [hunk 8]
```diff
@@ -9,9 +9,18 @@ <h1 *ngIf="title !== ''" data-role="transactions-view__title">
       </div>
       <ul *ngIf="transactions$ | async as items; else loading" class="bb-list">
         <li
-          *ngFor="let item of items | filterTransactions: filter"
+          *ngFor="
+            let item of items | filterTransactions: filter:selectedAccount
```
Comment (kirillsud) [inline libs/transactions/src/lib/views/transactions-view/transactions-view.component.html:13]: Isn't better to combine `transactions$` and `queryParams` Observables and do filtering in rxjs pipe?

**Review by kirillsud — CHANGES_REQUESTED** [review]

File: libs/transactions/src/lib/views/transactions-view/transactions-view.component.html (hunk 6 above)
Comment (Farzannajipour) [reply libs/transactions/src/lib/views/transactions-view/transactions-view.component.html:19]: And, based on the discussion we had, we can have a method to handle `id`.

File: libs/transactions/src/lib/services/transactions-details-route-resolver.service.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+@Injectable()
+export class TransactionsDetailsRouteResolverService
+  implements Resolve<TransactionItem | null>
+{
+  resolve(): Observable<TransactionItem> | Observable<null> {
+    const transactionData = this.router.getCurrentNavigation()?.extras
+      ?.state as TransactionItem;
+    if (!transactionData) {
+      return of(null);
```
Comment (kirillsud) [inline libs/transactions/src/lib/services/transactions-details-route-resolver.service.ts:14]: I was expecting loading data here

File: libs/transactions/src/lib/transactions-journey.module.ts [hunk 10]
```diff
@@ -17,21 +19,32 @@ import { TransactionsJourneyConfiguration } from './services/transactions-journe
...
+      title: TransactionsRouteTitleResolverService,
+    },
   },
-  resolve: {
-    title: TransactionsRouteTitleResolverService,
+  {
+    path: ':id',
+    component: TransactionDetailsComponent,
+    resolve: {
+      myData: TransactionsDetailsRouteResolverService,
```
Comment (kirillsud) [inline libs/transactions/src/lib/transactions-journey.module.ts:44]: Maybe call it `transactionData` or just `transaction`?

File: libs/transactions/src/lib/views/transaction-details/transaction-details-view.component.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  selector: 'bb-transaction-details',
+})
+export class TransactionDetailsComponent implements OnInit, OnDestroy {
+  public transactionDetails: ITransactionDetails | null = null;
+  private sub: Subscription = new Subscription();
+  constructor(public route: ActivatedRoute, private router: Router) {}
+
+  ngOnInit(): void {
+    this.sub.add(
+      this.route.data.subscribe(({ myData }) => {
```
Comment (kirillsud) [inline libs/transactions/src/lib/views/transaction-details/transaction-details-view.component.ts:22]: Why don't you use `snapshot`?

File: libs/transactions/src/lib/transactions-journey.module.ts (hunk 3 above)
Comment (manisha-backbase) [reply libs/transactions/src/lib/transactions-journey.module.ts]: done

File: libs/transactions/src/lib/views/transaction-details/transaction-details-view.component.html [hunk 12]
```diff
@@ -0,0 +1,19 @@
+<div class="row">
+  <div class="bb-card col-md-6">
+    <div class="bb-card__body">
+      <h1 data-role="transactions-view__title" class="bb-block--md">
+        Transaction Details
```
Comment (kirillsud) [inline libs/transactions/src/lib/views/transaction-details/transaction-details-view.component.html:5]: Missing translation

File: libs/transactions/src/lib/views/transaction-details/transaction-details-view.component.ts (hunk 11 above)
Comment (kirillsud) [reply libs/transactions/src/lib/views/transaction-details/transaction-details-view.component.ts:22]: Or `AsycPipe`?

File: libs/transactions/src/lib/pipes/filter-transactions.pipe.ts [hunk 13]
```diff
@@ -5,17 +5,24 @@ import { TransactionItem } from '@backbase/data-ang/transactions';
   name: 'filterTransactions',
 })
 export class FilterTransactionsPipe implements PipeTransform {
-  transform(value: TransactionItem[], text: string): TransactionItem[] {
-    if (!text) {
-      return value;
+  transform(
+    value: TransactionItem[],
+    text: string,
+    accountId?: string | null
```
Comment (kirillsud) [inline libs/transactions/src/lib/pipes/filter-transactions.pipe.ts:11]: Do we need it still?

File: libs/transactions/src/lib/views/transactions-view/transactions-view.component.ts (hunk 1 above)
Comment (manisha-backbase) [reply libs/transactions/src/lib/views/transactions-view/transactions-view.component.ts]: This code has changed.

File: libs/transactions/src/lib/views/transaction-details/transaction-details.component.ts (hunk 2 above)
Comment (manisha-backbase) [reply libs/transactions/src/lib/views/transaction-details/transaction-details.component.ts]: Added a resolver but for fetching the data from the API I have created a TODO task

File: libs/transactions/src/lib/views/transactions-view/transactions-view.component.ts [hunk 14]
```diff
@@ -15,23 +15,35 @@ import { TransactionsHttpService } from '../../services/transactions.http.servic
 export class TransactionsViewComponent {
   public title = this.route.snapshot.data['title'];
   public filter = '';
-
+  public bankAlias$ = this.route.queryParams;
```
Comment (kirillsud) [inline libs/transactions/src/lib/views/transactions-view/transactions-view.component.ts:18]: Maybe 
```suggestion
  public bankAlias$: Observable<string> = this.route.queryParams.pipe(map(x => x['bankAlias']);
```

File: libs/transactions/src/lib/views/transactions-view/transactions-view.component.html (hunk 8 above)
Comment (manisha-backbase) [reply libs/transactions/src/lib/views/transactions-view/transactions-view.component.html]: Done

File: libs/transactions/src/lib/views/transactions-view/transactions-view.component.html [hunk 15]
```diff
@@ -5,13 +5,28 @@ <h1 *ngIf="title !== ''" data-role="transactions-view__title">
...
+        *ngIf="(bankAlias$ | async)?.['bankAlias'] as accountName"
+      >
+        Filtered By:
+        <bb-badge-ui
+          [text]="accountName"
+          color="success"
+          [abbr]="false"
+          [circle]="false"
+        ></bb-badge-ui>
+        <span routerLink="/transactions">X</span>
```
Comment (kirillsud) [inline libs/transactions/src/lib/views/transactions-view/transactions-view.component.html:21]: In this way we also remove `search` query parameter :(

File: libs/transactions/src/lib/transactions-journey.module.ts (hunk 5 above)
Comment (manisha-backbase) [reply libs/transactions/src/lib/transactions-journey.module.ts:88]: But why?

File: libs/transactions/src/lib/views/transactions-view/transactions-view.component.ts [hunk 16]
```diff
@@ -15,23 +15,35 @@ import { TransactionsHttpService } from '../../services/transactions.http.servic
...
+    private router: Router,
     private readonly transactionsService: TransactionsHttpService,
     @Optional()
     @Inject(TRANSACTIONS_JOURNEY_COMMUNICATION_SERIVCE)
     private externalCommunicationService: TransactionsCommunicationService
   ) {}
+
+  search(ev: string) {
+    this.filter = ev || '';
+    this.router?.navigate([], { queryParams: { search: this.filter } });
```
Comment (kirillsud) [inline libs/transactions/src/lib/views/transactions-view/transactions-view.component.ts:47]: Don't forget about other current query parameters: `bankAlias` and `id`

**Review by kirillsud — CHANGES_REQUESTED** [review]

File: libs/transactions/src/lib/transactions-journey.module.ts (hunk 5 above)
Comment (manisha-backbase) [reply libs/transactions/src/lib/transactions-journey.module.ts:88]: defaultRoute is an array of routes now

Comment (kirillsud) [reply libs/transactions/src/lib/transactions-journey.module.ts:88]: the logic is that we use a custom `route` and have a fallback to `defaultRoute`. In your change we do the opposite therefore we always receive `defaultRoute` event if `route` is set

File: libs/transactions/src/lib/services/transactions-details-route-resolver.service.ts (hunk 9 above)
Comment (manisha-backbase) [reply libs/transactions/src/lib/services/transactions-details-route-resolver.service.ts:15]: I added a TODO task for this

File: libs/transactions/src/lib/views/transactions-view/transactions-view.component.html (hunk 15 above)
Comment (manisha-backbase) [reply libs/transactions/src/lib/views/transactions-view/transactions-view.component.html:21]: <span routerLink="/transactions" [queryParams]="{search: filter}">X</span>

Is this okii?

File: libs/transactions/src/lib/views/transactions-view/transactions-view.component.ts (hunk 16 above)
Comment (manisha-backbase) [reply libs/transactions/src/lib/views/transactions-view/transactions-view.component.ts:47]: I added queryParamsHandling: "merge"

File: libs/transactions/src/lib/services/transactions-details-route-resolver.service.ts (hunk 9 above)
Comment (kirillsud) [reply libs/transactions/src/lib/services/transactions-details-route-resolver.service.ts]: I've pushed that to the branch

File: libs/transactions/src/lib/views/transactions-view/transactions-view.component.html (hunk 15 above)
Comment (kirillsud) [reply libs/transactions/src/lib/views/transactions-view/transactions-view.component.html:21]: Perfect!

**Comment by kirillsud** [comment]
I've resolved all comments. Below you can find some notes about the updates.

### Communication inside journeys

1. Fixed query parameters update for both filters
2. Added error page for wrong transaction ID for Transaction Details page

image

### Communication between journeys

1. Remove account name query parameters for *View Transactions* link: `/transactions?account=424b4167-6dcc-41ed-92cf-5d5e472f64fd`

### Usability Improvement

1. Improved UI for accounts filter badge
2. Improved UI for transaction details page
3. Added `<a>` tags for all `[routerLink]` directives to make links more visible for browsers

image

image

File: apps/golden-sample-app/src/app/app.error-handler.ts [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+    private readonly router: Router,
+    private readonly ngZone: NgZone
+  ) {}
+
+  handleError(error: Error | unknown | { rejection: Error | unknown }): void {
+    const errorObject = toErrorObject(error);
+
+    if (errorObject?.name === 'NotFoundError') {
+      this.ngZone.run(() => {
+        this.router.navigate(['/error'], {
```
Comment (kirillsud) [inline apps/golden-sample-app/src/app/app.error-handler.ts:33]: What do you think about this way of handling page errors which should be rendered as an error page?

**Review by kirillsud — APPROVED** [review]

**Review by santam85 — APPROVED** [review]

**Comment by santam85** [comment]
@manisha-backbase @kirillsud very nice! Can you check the typo and the couple of minor type warnings ? Some of them are also on unchanged files but they are flagged in the PR anyway. If we do that I think it's good to merge this!

File: apps/golden-sample-app/src/app/app.error-handler.ts (hunk 17 above)
Comment (santam85) [reply apps/golden-sample-app/src/app/app.error-handler.ts:33]: I think it's a nice example but probably not very feasible at scale as it implies having all error UIs in a central place.
We can leave it like this, but I'd prefer if we just handle the 404 route with the https://designsystem.backbase.com/v2/web-components/common-error-state/web component and delegate all the rest of the error handling isolating it in the journey libs. Its a pattern that scales better with the size of the team we have.

**Review by kirillsud — APPROVED** [review]

**Pull request merged**